### PR TITLE
Modify the add logic of the render window

### PR DESCRIPTION
**Pull request by GengineJS**

### CODE REVIEW

**Comment by dumganhar** [comment]
Does `Root.cpp` also need to be updated?

**Comment by GengineJS** [comment]
> Does `Root.cpp` also need to be updated?

Yes, I'll fix it right away

File: cocos/core/renderer/core/render-window.ts [hunk 1]
```diff
@@ -256,17 +261,25 @@ export class RenderWindow {
...
+                if (!this._cameras.length) this._detachFromRoot();
                 return;
             }
         }
     }
 
+    private _detachFromRoot () {
+        const windows = legacyCC.director.root.windows;
+        const idx = windows.indexOf(this);
+        if (idx !== -1) windows.splice(idx, 1);
```
Comment (minggo) [inline cocos/core/renderer/core/render-window.ts]: Modify windows outside of root is not a good idea. It is better modify it in Root.

Comment (GengineJS) [reply cocos/core/renderer/core/render-window.ts]: have modified

File: cocos/core/root.ts [hunk 2]
```diff
@@ -535,15 +534,34 @@ export class Root {
...
+            this._windows.push(window);
+        }
+    }
+
+    /**
+    * @en Detach a render window
+    * @zh 移除指定的窗口
+    * @param window The render window to be detached
+    */
+    public detachWindow (window: RenderWindow) {
```
Comment (minggo) [inline cocos/core/root.ts:559]: detachWindow -> removeWindow?

File: cocos/core/root.ts [hunk 3]
```diff
@@ -535,15 +534,34 @@ export class Root {
...
+            this.detachWindow(window);
         }
     }
 
+    /**
+    * @en Attach a render window
+    * @zh 添加指定的窗口
+    * @param window The render window to be attached
+    */
+    public attachWindow (window: RenderWindow) {
```
Comment (minggo) [inline cocos/core/root.ts:548]: attachWindow -> addWindow?

File: cocos/core/root.ts [hunk 4]
```diff
@@ -525,7 +525,6 @@ export class Root {
     public createWindow (info: IRenderWindowInfo): RenderWindow | null {
         const window = this._createWindowFun(this);
         window.initialize(this.device, info);
-        this._windows.push(window);
```
Comment (minggo) [inline cocos/core/root.ts:528]: Why remove it?

File: cocos/core/renderer/core/render-window.ts [hunk 5]
```diff
@@ -256,17 +258,23 @@ export class RenderWindow {
         for (let i = 0; i < this._cameras.length; ++i) {
             if (this._cameras[i] === camera) {
                 this._cameras.splice(i, 1);
+                if (!this._cameras.length) this._detachFromRoot();
                 return;
             }
         }
     }
 
+    private _detachFromRoot () {
```
Comment (minggo) [inline cocos/core/renderer/core/render-window.ts:267]: _detachFromRoot -> removeFromWindow？

File: cocos/core/root.ts (hunk 4 above)
Comment (GengineJS) [reply cocos/core/root.ts:528]: The logic of whether to add a window depends on whether the camera is used or not. If the camera is not used, it is not necessary to add renderWindow to the windows property of the root

Comment (minggo) [reply cocos/core/root.ts:528]: Got it. Sounds reasonable.

**Review by minggo — APPROVED** [review]

**Comment by minggo** [comment]
@cocos-robot run test cases

**Review by dumganhar — APPROVED** [review]

**Review by star-e — APPROVED** [review]

**Comment by star-e** [comment]
@cocos-robot run test cases

**Comment by GengineJS** [comment]
@cocos-robot run test cases

**Comment by GengineJS** [comment]
@cocos-robot run test cases

**Comment by GengineJS** [comment]
@cocos-robot run test cases

**Comment by GengineJS** [comment]
@cocos-robot run test cases

**Comment by GengineJS** [comment]
@cocos-robot run test cases

**Comment by GengineJS** [comment]
@cocos-robot run test cases

**Comment by GengineJS** [comment]
@cocos-robot run test cases

**Comment by GengineJS** [comment]
@cocos-robot run test cases

**Comment by GengineJS** [comment]
@cocos-robot run test cases

**Review by star-e — APPROVED** [review]

**Comment by GengineJS** [comment]
@cocos-robot run test cases

**Comment by dumganhar** [comment]
@cocos-robot run test cases

**Pull request closed without merging**